File: data structures/heap/Heap.java
import java.util.Arrays;

// min heap
public class Heap {
    private int[] heap;
    private int size;

    public Heap() {
        this(10);
    }

    public Heap(int size) {
        heap = new int[size + 1];
        this.size = size;
    }

    public void push(int item) {
        heap[++size] = item;
        int index = size;

        while(index != 1 && item < heap[index / 2]) {
            heap[index] = heap[index / 2];
            index /= 2;
        }

        heap[index] = item;
    }

    public int pop() {
        int item = heap[1];
        int last = heap[size--];
        int parent = 1, child = 2;

        while(child <= size) {
            if(child < size && heap[child + 1] < heap[child]) {
                child++;
            }
            if(last <= heap[child])
                break;

            heap[parent] = heap[child];
            parent = child;
            child *= 2;
        }

        heap[parent] = last;

        return item;
    }

    @Override
    public String toString() {
        return Arrays.toString(Arrays.copyOfRange(heap, 1, size + 1));
    }
}

File: data structures/heap/heap.py
# min heap
class Heap:

    def __init__(self) -> None:
        self.heap = [None]
        self.size = 0
    
    def push(self, item) -> None:
        self.size += 1
        if len(self.heap) <= self.size:
            self.heap.append(None)

        index = self.size
        while index != 1 and item < self.heap[index // 2]:
            self.heap[index] = self.heap[index // 2]
            index //= 2
        
        self.heap[index] = item
    
    def pop(self):
        item = self.heap[1]
        last = self.heap[self.size]
        self.size -= 1
        parent = 1
        child = 2

        while child <= self.size:
            if child < self.size and self.heap[child + 1] < self.heap[child]:
                child += 1
            
            if last <= self.heap[child]:
                break

            self.heap[parent] = self.heap[child]
            parent = child
            child *= 2
        
        self.heap[parent] = last
        
        return item

    def __str__(self) -> str:
        return "[" + " ".join(map(str, self.heap[1:self.size+1])) + "]"